subsys/pm/pm.c

```diff
@@ -182,40 +185,43 @@ void pm_system_resume(void)
 
 bool pm_state_force(uint8_t cpu, const struct pm_state_info *info)
 {
-	bool ret = false;
+	k_spinlock_key_t key;
 
 	__ASSERT(info->state < PM_STATE_COUNT,
 		 "Invalid power state %d!", info->state);
 
+	key = k_spin_lock(&pm_forced_state_lock);
+	z_cpus_pm_forced_state[cpu] = *info;
+	k_spin_unlock(&pm_forced_state_lock, key);
 
-	if (!atomic_test_and_set_bit(z_cpus_pm_state_forced, cpu)) {
-		z_cpus_pm_state[cpu] = *info;
-		ret = true;
-	}
-
-	return ret;
+	return true;
 }
 
 bool pm_system_suspend(int32_t ticks)
 {
 	uint8_t id = _current_cpu->id;
+	k_spinlock_key_t key;
 
 	SYS_PORT_TRACING_FUNC_ENTER(pm, system_suspend, ticks);
 
-	if (!atomic_test_bit(z_cpus_pm_state_forced, id)) {
+	key = k_spin_lock(&pm_forced_state_lock);
+	if (z_cpus_pm_forced_state[id].state != PM_STATE_ACTIVE) {
```

Comment:
For the purpose of this pm_forced_state check we probably should have something like PM_STATE_NONE or use other field that will indicate if there is any forced_state requested. Using PM_STATE_ACTIVE looks like a shortcut that might be a bit confusing. Are you confident that no one will ever attempt to force active state to resume from some deep suspend state?

Reply:
the force state does not wake up the cpu from a sleep state. This only used next time the system is idle, if you are setting it to ACTIVE the system will ignore it and ask the policy which state should be used, if one really wants to keep the cpu active it should use constraints. Though, having a invalid state may make things clear indeed. This change though does not need to be done right now IMHO.

Reply:
@ceolin I am aware that this implementation will most likely work fine. However, while your explanation makes sense it seem to be a misuse of a PM_STATE_ACTIVE. I would recommend to clean this along with race condition fix that you provide.

Reply:
Add a new `state` does not look appropriated to me, what a `NONE` state means ? Maybe have something that is used only internally is the way to go.

Reply:
> Maybe have something that is used only internally is the way to go.

@ceolin what do you mean by that? I would say that we need a clean way to determine if force state is requested or not. If not by additional `NONE` state (that would in that case indicate that state field is not set) then by some other field that will be only set when pm_state_force request is received.

Reply:
One additional state is cleaner IMHO, since there is one less thing to keep sync. The problem is that states are public interface and all of them represent a clear power state. NONE does not mean anything ... What I said is that we may keep this internally.

Reply:
> One additional state is cleaner IMHO, since there is one less thing to keep sync. The problem is that states are public interface and all of them represent a clear power state. NONE does not mean anything ... What I said is that we may keep this internally.

I understand your concern, but the pm state force support is public as well, right? That is why I suggest to have a clean solution on upstream. I don't need (and do not want) any internal only change for that. 

Anyway, that is just my feedback. It is up to  you @ceolin and maintainers how you want to proceed. At the end I want to have working pm force state support.

Reply:
Let me see if I get it better now, in this case the `NONE` state would also be used to cancel a previous forced state ?
for example:
```
...
pm_state_force(PM_STATE_SOFT_OFF);
....
/* something happened and the application no longer wants to suspend */
pm_state_force(PM_STATE_NONE);
....
```

Is that the case ?

Reply:
That is what I meant:
```
@@ -25,18 +25,21 @@ LOG_MODULE_REGISTER(pm, CONFIG_PM_LOG_LEVEL);
 static ATOMIC_DEFINE(z_post_ops_required, CONFIG_MP_NUM_CPUS);
 static sys_slist_t pm_notifiers = SYS_SLIST_STATIC_INIT(&pm_notifiers);
 
+
+#define PM_STATE_NONE -1
+
 /*
  * Properly initialize cpu power states. Do not make assumptions that
  * ACTIVE_STATE is 0
  */
-#define CPU_PM_STATE_INIT(_, __)		\
-	{ .state = PM_STATE_ACTIVE }
+#define CPU_PM_STATE_INIT(_, _state)		\
+	{ .state = (_state) }
 static struct pm_state_info z_cpus_pm_state[] = {
-	LISTIFY(CONFIG_MP_NUM_CPUS, CPU_PM_STATE_INIT, (,))
+	LISTIFY(CONFIG_MP_NUM_CPUS, CPU_PM_STATE_INIT, (,), PM_STATE_ACTIVE)
 };
 
 static struct pm_state_info z_cpus_pm_forced_state[] = {
-	LISTIFY(CONFIG_MP_NUM_CPUS, CPU_PM_STATE_INIT, (,))
+	LISTIFY(CONFIG_MP_NUM_CPUS, CPU_PM_STATE_INIT, (,), PM_STATE_NONE)
 };
 
 static struct k_spinlock pm_forced_state_lock;
@@ -205,9 +208,9 @@ bool pm_system_suspend(int32_t ticks)
 	SYS_PORT_TRACING_FUNC_ENTER(pm, system_suspend, ticks);
 
 	key = k_spin_lock(&pm_forced_state_lock);
-	if (z_cpus_pm_forced_state[id].state != PM_STATE_ACTIVE) {
+	if (z_cpus_pm_forced_state[id].state != PM_STATE_NONE) {
 		z_cpus_pm_state[id] = z_cpus_pm_forced_state[id];
-		z_cpus_pm_forced_state[id].state = PM_STATE_ACTIVE;
+		z_cpus_pm_forced_state[id].state = PM_STATE_NONE;
 	} else {
 		const struct pm_state_info *info;
 
```

Reply:
@ceolin that is the implementation I was thinking of, this check makes much more sense to me:
`if (z_cpus_pm_forced_state[id].state != PM_STATE_NONE)`

That way when no pm state force is requested the field is reset to PM_STATE_NONE. However I am not expecting any scenario where application would attempt to actually cancel forced pm state by writing none.